docs/architecture/adr-002-lite-client.md

```diff
@@ -0,0 +1,454 @@
+future commit.
+
+### Full Verification
+
+Since we know the validators for a commit 
+(ie. the number of validators should match the length of our votes vector), 
+we can iterate over them, check the signatures, and sum the voting power.
+
+An error should be returned if:
+
+- any signature is invalid 
+- 2/3 or less of the voting power signed
+
+So we can have a function like:
+
+```rust
+fn verify_commit_full<V, C>(vals: V, commit: C) -> Result<(), Error>
+where
+    V: ValidatorSet,
+    C: Commit,
+{
+```
+
+### "Trusting" Verification
+
+To do skipping verification (ie. the "trusting method"), 
+we have to check if +1/3 of validators at some past height signed the commit, 
+before we can check if +2/3 of the validators for the current height signed.
+To do this, we have to know which validator a vote is from and be able to look them up in the validator
+set. Hence we need to extend our ValidatorSet trait to permit such lookups:
+
+```rust
+pub trait ValidatorSetLookup: ValidatorSet {
+    fn validator(&self, val_id: ValID) -> Option<Self::Validator>;
+}
+```
+
+Now for each vote in the commit, we can check if the validator existed in our
+trusted validator set, and thus if +1/3 of the trusted validators signed the new
+commit. We can use a function with the same signature:
+
+```rust
+fn verify_commit_trusting<V, C>(vals: V, commit: C) -> Result<(), Error>
+where
+    V: ValidatorSet,
+    C: Commit,
+{
+```
+
+### Validation
+
+Most of the above is about checking the signatures and voting power in commits, 
+but we also need to perform other validation checks, 
+like that the validator set hashes match what are in the header, and the lite
+client's trusted state actually hasn't expired. Pure functions for all of these
+checks should be provided.
+
+Some things are left explicitly unvalidated as they have minimal bearing on the correctness of the lite client.
+These include:
+
+- LastCommitHash
+	- In the skipping case, it's not possible to verify the header refers to the correct previous block without reverting to the sequential case. So in the sequential case, we don't validate this either. If it's incorrect, in indicates the validators are misbehaving, though we can only detect it as a lite client if there's a fork.
+- BlockID
+	- As mentioned, this includes a merkle root of the entire block and is not verifiable without downloading the whole block, which would defeat the purpose of the lite client!
+- Time
+	- Verifying the time would require us to download the commit for the previous block, and to take the median of the timestamps from that commit. This would add significant overhead to the lite client (an extra commit to validate for every block!). If the time is incorrect, in indicates that the validators are explicitly violating the protocol rules in a detectable way which full nodes should detect in the first place and shouldn't forward to lite clients, so there would probably be bigger issues at foot.
```

Comment:
```suggestion
	- Verifying the time would require us to download the commit for the previous block, and to take the median of the timestamps from that commit. This would add significant overhead to the lite client (an extra commit to validate for every block!). If the time is incorrect, it indicates that the validators are explicitly violating the protocol rules in a detectable way which full nodes should detect in the first place and shouldn't forward to lite clients, so there would probably be bigger issues at foot.
```